docs/content/develop/permadiff.md

```diff
@@ -0,0 +1,261 @@
+{{< /tabs >}}
+
+## API does not return a field that was sent
+
+In the flattener for the field, return the value of the field in the user's configuration.
+
+{{< tabs "ignore_read" >}}
+{{< tab "MMv1" >}}
+On top-level fields, this can be done with:
+
+```yaml
+ignore_read: true
+```
+
+For nested fields, `ignore_read` is [not currently supported](https://github.com/hashicorp/terraform-provider-google/issues/12410), so this must be implemented with a [custom flattener]({{< ref "/develop/custom-code#custom_flatten" >}}). You will also need to add the field to `ignore_read_extra` on any examples that are used to generate tests; this will cause tests to ignore the field when checking that the values in the API match the user's configuration.
+
+```go
+func flatten<%= prefix -%><%= titlelize_property(property) -%>(v interface{}, d *schema.ResourceData, config *transport_tpg.Config) interface{} {
+    // We want to ignore read on this field, but cannot because it is nested
+    return d.Get("path.0.to.0.nested.0.field")
+}
+```
+
+```yaml
+examples:
+ - !ruby/object:Provider::Terraform::Examples
+   # example configuration
+   ignore_read_extra:
+     - "path.0.to.0.nested.0.field"
+```
+{{< /tab >}}
+{{< tab "Handwritten" >}}
+Use `d.Get` to set the flattened value to be the same as the user-configured value (instead of a value from the API).
+
+```go
+func flattenParentField(d *schema.ResourceData, disk *compute.AttachedDisk, config *transport_tpg.Config) []map[string]interface{} {
+    result := map[string]interface{}{
+        "nested_field": d.Get("path.0.to.0.parent_field.0.nested_field")
+    }
+    return []map[string]interface{}{result}
+}
+```
+
+In tests, add the field to `ImportStateVerifyIgnore` on any relevant import steps.
+
+```
+{
+    ResourceName:            "google_product_resource.default",
+    ImportState:             true,
+    ImportStateVerify:       true,
+    ImportStateVerifyIgnore: []string{""path.0.to.0.parent_field.0.nested_field"},
+},
+```
+{{< /tab >}}
+{{< /tabs >}}
+
+## API returns a list in a different order than was sent
+
+For an Array of nested objects, convert it to a Set – this is a [breaking change]({{< ref "/develop/breaking-changes/breaking-changes" >}}) and can only happen in a major release.
+
+For an Array of simple values (such as strings or ints), rewrite the value in the flattener to match the order in the user's configuration. This will also simplify diffs if new values are added or removed.
+
+{{< tabs "diff_suppress_list" >}}
+
+{{< tab "MMv1" >}}
+Add a [custom flattener]({{< ref "/develop/custom-code#custom_flatten" >}}) for the field.
+
+```go
+func flatten<%= prefix -%><%= titlelize_property(property) -%>(v interface{}, d *schema.ResourceData, config *transport_tpg.Config) interface{} {
+    configValue := d.Get("path.0.to.0.parent_field.0.nested_field").([]string)
+
+    ret := []string{}
+    // Add values from v to ret to match order in configValue and put any new strings at the end
```

Comment:
I couldn't find a reference implementation for this and didn't want to write code that's not tested; we could potentially update this with the results from https://github.com/hashicorp/terraform-provider-google/issues/14591 (especially if that ends up providing a utility function that could be called here.)

Reply:
A utility function for simple slices would be the ideal solution here. Especially as this custom flattener would need to consider not only reordering the values from the API to match the config, but also whether there are absent or extra values on client or API side in the context of that reordering process. It can get a bit hairy!

Reply:
attn @sahsagar-google FYI